ENH, MAINT: Refactor `PyArray_InnerProduct` to use `PyArray_MatrixProduct2`

jakirkham opened the pull request:
Fixes https://github.com/numpy/numpy/issues/6948

Related: https://github.com/numpy/numpy/pull/6932
Related: https://github.com/numpy/numpy/pull/6977
Related: https://github.com/numpy/numpy/pull/6986
Related: https://github.com/numpy/numpy/pull/6987
Related: https://github.com/numpy/numpy/pull/6988

This follows @seberg's suggestion ( https://github.com/numpy/numpy/issues/6948#issuecomment-169769213 ) and does the simplest thing. Namely, refactors `PyArray_InnerProduct` to use `PyArray_MatrixProduct2`. As a consequence, `np.inner` will see a speedup in cases where the problem contains portions like these `a @ a.T` and `a.T @ a` as this already is optimized in `cblas_matrixproduct` because of this PR ( https://github.com/numpy/numpy/pull/6932 ).

<br>
Todo:
- [x] Refactor so that `PyArray_InnerProduct` just calls `PyArray_MatrixProduct2`.
- [x] Add more tests of different dimensions for `np.inner` to make sure it still behaves correctly.
- [x] Add benchmarks so that cases where `np.inner` can now see a speedup are shown vs those it can't as was done with `np.dot`.
- [x] Add test for type mismatch exception.

jakirkham committed:
MAINT: Refactor `cblas_innerproduct` to use `cblas_matrixproduct`.

seberg commented:
Yeah, that is what I mean. I will let someone else decide further, but I think the transpose may need to be a bit more complicated. To figure that out/test it, can you add a test for multiplying larger then 2D arrays?

jakirkham commented:
My understanding is that even from the C-API this isn't intended for public consumption as it only works on 2D arrays or smaller just as `cblas_matrixproduct`. Instead, there is a `PyArray_InnerProduct`, which is intend for use in the C API and is bound to `np.inner` in the Python API.

This only gets called in one place ( https://github.com/numpy/numpy/blob/eeba2cbfa4c56447e36aad6d97e323ecfbdade56/numpy/core/src/multiarray/multiarraymodule.c#L848-L854 ). As you can see, we are ensured to only have arrays with dimension 2 or less inside `cblas_innerproduct`.

seberg commented:
Ah, right. Frankly, I would much prefer if we can do it for PyArray_InnerProduct making it just call PyArray_MatrixProduct2. Then remove the whole cblas_innerproduct function. Though I am not sure it makes sense, my guess would be it does.

jakirkham commented:
It looks like there are some `einsum` tests that also given `inner` 3D arrays, but I don't know that this is the best way to test `inner` on these cases by comparing it to the much more complex `einsum`. I can certainly add some. ( https://github.com/jakirkham/numpy/blob/e72e1510d892ce4464cf102000e89582327953a0/numpy/core/tests/test_einsum.py#L262-L275 ).

Other things this probably needs are some benchmarks as I am claiming one can now get a speedup with `inner`.

jakirkham commented:
> Ah, right. Frankly, I would much prefer if we can do it for PyArray_InnerProduct making it just call PyArray_MatrixProduct2. Then remove the whole cblas_innerproduct function. Though I am not sure it makes sense, my guess would be it does.

It is probably removable, but what I liked about your original proposal (at least how I read it) is everything stays pretty much the same in terms of how everything builds. Also, how many, which, and where functions are remains the same. As soon as we start messing with that, we open ourselves up to spending time hunting down potentially weird build errors. At present, this just works.

seberg commented:
Well, look at MatrixProduct2 ;), it is a monster of annoying stuff and all that is needed to call it from inner is transposing the last two axes of op2 (if op2.ndim >= 2). The added complexity is only that the transpose is more complicated because you cannot pass in NULL. Other then that, you just remove more complexity. Note that the function you changed can be removed completely, it is not public.

jakirkham commented:
It would be nice if there was a `PyArray_RollAxis`.

jakirkham commented:
So, why is that we have `PyArray_MatrixProduct` and `PyArray_MatrixProduct2` if they just duplicate functionality that is part of `inner`.

seberg commented:
Well, I don't know, you probably have to go way back to figure that one out ;). That is why I would like it refactored away. You save a single tranpose (array creation) for all I figure.

jakirkham commented:
Sorry, I realized I was asking a question that basically became "why do we have `numpy.dot` and `numpy.inner`?" and decided I should just be quiet least this gets derailed.

jakirkham commented:
Btw, did you see this ( https://github.com/numpy/numpy/issues/5859 )?

jakirkham committed:
MAINT: Refactor `PyArray_InnerProduct` so that it just performs a transpose and calls `PyArray_MatrixProduct2`.

jakirkham commented:
So, if I make this simple change ( https://github.com/jakirkham/numpy/commit/d2ea634cfd7413ee784cb56d826a2a4167e903c3 ), I get a segmentation fault in the test suite. Is there something that I am doing wrong here? I'm not very familiar with the C API so I wouldn't be surprised if I am. I just need a few pointers.

seberg commented:
Can you do me the favor and try to refactor all of PyArray_Inner? At least to me doing the transpose specifically in this subfunction seems half baked. Doing the full transpose should not be too difficult, see for example https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/mapping.c#L135 for a "complex" Transpose operation, but it gives you the idea that you can just iterate and set all ndim (if ndim >= 2), and then switch the last two ndim around.

jakirkham commented:
So, I actually did that too. Though I have pushed it to the PR yet. I am struggling with a segmentation fault there too. Here is the commit ( https://github.com/jakirkham/numpy/commit/8f5464ce59772bce870d38b4b50dc282e0cde5c1 ).

jakirkham commented:
Ok, so I have pushed the refactoring of `PyArray_InnerProduct` using `PyArray_MatrixProduct2` to the PR, @seberg. I got rid of the decrefs for `op*`, but it is still failing. However, now it is a reference count error. Here is one of the builds on my end for an example. ( https://travis-ci.org/jakirkham/numpy/jobs/101126350 )

jakirkham commented:
Actually, I think I found a couple typos that may have caused some issues. Will retest and see if I am still having problems.

jakirkham commented:
Weird. So, I don't need to use decref on a `PyArray_Descr` instance if it is given to `PyArray_FromAny`. I would not have expected that.

jakirkham commented:
In any event, we are not getting segfaults or other weird errors any more. It is just a matter of making sure the tests are passing now, which I know they are not.

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -813,121 +813,52 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
-        dimensions[j++] = PyArray_DIMS(ap1)[i];
+    newaxes.ptr = dims;
+    newaxes.len = PyArray_NDIM(ap2);
+    dims[0] = newaxes.len - 1;
```
It seems you are swapping to the first position, You have to switch the last two positions.

jakirkham replied:
So, I am basing this off the sample code that I found to work yesterday. ( https://github.com/numpy/numpy/issues/6948#issuecomment-169733144 ) I think I am doing the transpose correctly, but I did a reshape before, which I have not done yet. Unless there is a way to avoid the reshape completely.

seberg replied:
I am pretty sure `inner` is equivalent to `np.dot(a, np.swapaxes(b, -1, -2))`.

jakirkham replied:
Ok, I think you are correct. I think I was just trying to hold as closely to the `np.tensordot(a, b, axes=(-1,-1))[()]` based definition as I could. I'll change it.

jakirkham replied:
I think part of the reason I went away from that is if you had fewer dimensions than `2`. So, I need to fix up my logic for that case.

jakirkham replied:
There was also another special case I forgot about. Handling `np.inner(scalar, 2d_array)`. In this case, we don't want to transpose.

jakirkham commented:
Alright, I think we are covering the right sorts of cases now. Any thoughts on how it is looking thus far, @seberg?

njsmith commented:
Yeah, most functions that take PyArray_Descr arguments steal the reference
to it, i.e. if you want to hold on to the object afterwards, you must call
Py_INCREF before calling the function. I think this has to do with the
original Numeric->numpy transition, when PyArray_Descr got promoted from a
regular c struct to a full python object. It's confusing and error-prone,
but too late to do much about it...
On Jan 8, 2016 10:54 AM, "jakirkham" <notifications@github.com> wrote:

> Weird. So, I don't need to use decref on a PyArray_Descr instance if it
> is given to PyArray_FromAny. I would not have expected that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6968#issuecomment-170090622>.
>

jakirkham commented:
Alright, I think this is about done and could really use some feedback. On the C side, `inner` uses `dot` with a transpose beforehand. There is a benchmark that show that `inner` now sees the `syrk` speedup. I haven't added any new tests, but the exist test suite caught all the corner cases I expected before I implemented them in the refactored code.

Does this look ok? Are you happy with this approach? Are there still some changes needed? Are there some more tests I should have here? If you have any feedback, please let me know.

ping: @njsmith @seberg @pv @mhvk

mhvk commented:
My C is too poor to do a detailed review, but I very much like how much code has been removed! Out of curiosity, for a non-syrk case, is the timing essentially the same as before?

jakirkham commented:
On my computer, I ran the non-`syrk` benchmark case and they look basically the same around. Both were somewhere in the range of `570us` - `580us`. The old version might be a couple `us` lower typically, but it is hard to say as that digit appears uncertain anyways.

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -813,121 +813,72 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
     typec = PyArray_DescrFromType(typenum);
     if (typec == NULL) {
-        return NULL;
+        PyErr_SetString(PyExc_ValueError, "Cannot find a common data type.");
```
If you set the string here, does it also have to be set in MatrixProduct?

jakirkham replied:
Yeah, `PyArray_MatrixProduct2` is used by `np.dot` and the method `dot`, which doesn't touch this code. I suppose we could drop setting the string here, but we still have to handle the error. So, then we would just have an error with an unclear description.

seberg replied:
Nono, this is good, I just did not realize that this again was a forgotten change to `inner`....

Btw. could add a test for it with something like `np.dot(1., np.array((1,1), dtype='i,i'))`.

jakirkham replied:
Yeah, it was. So, I figured we should add it now that we are rewriting it.

Ok, I can add some tests for `np.dot` and `np.inner`.

FYI, I tried using `inner` as is now and this is there error I get. It is better than I thought it would, but this is still an improvement.

    >>> import numpy as np
    >>> np.inner(1., np.array((1,1), dtype='i,i'))
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: invalid type promotion

jakirkham replied:
So, I tried the following with and without are changes and got the same result in either case. In short, it never hits our error message. Also, it gives two different error messages depending on the order.

    >>> import numpy
    >>> import numpy as np
    >>> np.inner(1., np.array((1,1), dtype='i,i'))
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: invalid type promotion
    >>> np.inner(np.array((1,1), dtype='i,i'), 1.)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('float64') according to the rule 'safe'

seberg replied:
I am seeing your "Cannot cast to common type" message for the first example (first array is valid type). Anyway, I think this is good as is.

jakirkham replied:
Nevermind, I see it now too. I think I had some build detritus that messed everything up. Why is that we can't catch the second one though? Does `PyArray_ObjectType` do something weird?

seberg replied:
I think the casting ObjecType logic might be saying something like "yeah, you can cast both to void type". In any case the error in that case originates only from the FromAny below which cannot do the cast.

jakirkham replied:
Ah, ok, that makes more sense. The other question I have here is `dot` is throwing a `ValueError` and we are throwing a `TypeError`. Should they both be `TypeError`s or `ValueError`s?

jakirkham replied:
I have added some tests assuming the current behavior is correct.

jakirkham replied:
I wanted to point out that this a change in behavior as it raised a `TypeError` before. Anyways, I have put the test and this minor change in a separate PR ( https://github.com/numpy/numpy/pull/6987 ) that is merged into this one. That way, if we need to discuss this further, it can occur there.

seberg replied:
Well, on 1.8 this seemed to have caused segfault, so frankly I don't think we have to worry about changing the error type. On the safe side I would tend to not change `np.dot` since it is more used, but really don't know what is the better message.

jakirkham replied:
We can always drop the exception string change line above and it will preserve the same behavior if keeping it working the same way is a concern or we can just change it back to a `TypeError`, but leave the new description.

jakirkham replied:
So, while I like the idea of testing these. It results in a weird Python failure. Apparently it doesn't expect to have an exception at some point, but it still has one so this causes a crash in the test suite. Personally, I am a bit lost as to what is going on and could use some pointers.

jakirkham replied:
I have placed the tests in this PR ( https://github.com/numpy/numpy/pull/6988 ) so they can be looked at separately. Note that I am not changing the current behavior in that PR only testing it as it is.

jakirkham replied:
I decided that I probably should drop this exception line as it seems somewhat controversial and I am having difficulties getting simple tests to work for the exception.

jakirkham replied:
As I have fixed the exception tests, I have brought them back into this PR, but I have left the `ValueError` change out.

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -813,121 +813,72 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
-        dimensions[j++] = PyArray_DIMS(ap1)[i];
+    newaxes.len = PyArray_NDIM(ap2);
+    if ((PyArray_NDIM(ap1) >= 1) && (newaxes.len >= 2))
+    {
```
We put the { on the previous line, except for functions.

seberg replied:
Did not finish going through the code, will do it later.

jakirkham replied:
Ok, sorry about that. It's now corrected. Sure, let me know.

seberg commented:
@mhvk, I guess it might be noticable for smaller arrays...

seberg commented:
Looks good to me, did you check there are tests for 3-dim arrays, or could you add them if they don't exist? (The last `goto fail` is unnecessary, whatever you like)
@charris, I think I like this, even if the transpose might make things a bit slower for small arrays. Any opinions?

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -813,121 +813,70 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
-     * -- use priority to determine which subtype.
-     */
-    ret = new_array_for_sum(ap1, ap2, NULL, nd, dimensions, typenum);
+    ret = PyArray_MatrixProduct2(op1, ap2t, NULL);
```
Oh, please use ap1 here, if we cannot refactor it away, lets use it, since it might save an array creation or cast.

jakirkham replied:
Good call. I must have missed that. Now is `ap1`.

jakirkham commented:
> Looks good to me, did you check there are tests for 3-dim arrays, or could you add them if they don't exist?

There are some, but they are comparing `einsum` to `inner`. Though we probably should have some tests for `inner` that don't rely on another function (especially a more complex one) to determine success, which I am adding. Also, probably could use at least one `syrk` vs. `gemm` test.

> The last `goto fail` is unnecessary, whatever you like

For that matter `Py_XDECREF(ret);` is unnecessary to. I guess we could drop these and maybe it goes slightly faster, but I bet it's a wash. On the other hand, if someone goes back and adds code after this point, they won't have to remember to add any of this stuff as it is already there. If you feel strongly about it, I will remove it.

jakirkham committed:
TST: Ensure `dot` fails correctly if array types cannot be coerced into a common type.

jakirkham committed:
TST: Ensure `inner` fails correctly if array types cannot be coerced into a common type.

jakirkham commented:
Tried to add some tests for the exception, but they fail `USE_DEBUG=1` only due to some issue with the exception. Not really, sure what is going on there. Any pointers anybody has would be helpful.

jakirkham commented:
Maybe related to this ( https://github.com/numpy/numpy/issues/6741 ).

jakirkham commented:
So, I have removed the changed exception as it appears to be somewhat controversial and it is in a different PR ( https://github.com/numpy/numpy/pull/6987 ) with exception tests. I have also removed the exceptions tests as they seem to have issues that at the moment that I cannot seem to figure out and placed them in another PR ( https://github.com/numpy/numpy/pull/6988 ). If we can get these working, I am willing to combine them back into this PR. However, I don't want this PRs fate to be determined by trying to come up with acceptable solutions to these more minor issues.

jakirkham commented:
> Looks good to me, did you check there are tests for 3-dim arrays, or could you add them if they don't exist?

Yes, they went in this PR ( https://github.com/numpy/numpy/pull/6986 ), which is already merged. This PR has been rebased on master after those commits were added so includes them.

jakirkham committed:
BUG: Clear error before constructing error message using calls to `PyObject_Repr`. Also, do a better job of handling any errors raised while constructing the error message.

jakirkham commented:
Figured out the issues with type exceptions thanks to help from @seberg and @charris. As they are now working, I have brought that PR ( https://github.com/numpy/numpy/pull/6988 ) back into the history of this PR.

jakirkham commented:
Failures on AppVeyor ( https://github.com/numpy/numpy/issues/6991 ) had not previously occurred for this PR with exception of the segmentation fault issue (also happened on Travis), which has since been resolved. I believe these to be unrelated to the content of this PR.

jakirkham commented:
As AppVeyor is merging with master, which is broken, it is currently failing the tests there. So, I ran my own AppVeyor build without this merge (this is rebased on a commit on `master` before the bad commit) and it looks like it checks out ( https://ci.appveyor.com/project/jakirkham/numpy/build/1.0.4 ).

charris commented:
Thanks @jakirkham . The `cblasfuncs.c` could use some style fixes, but that isn't new here.

seberg commented:
Thanks!

jakirkham commented:
Thanks everyone.

Alright, @charris, I'll try to look at this at some point soon.

jakirkham commented:
The benchmark shows the speedup. ( https://pv.github.io/numpy-bench/#bench_linalg.Eindot.time_inner_trans_a_a )

mhvk commented:
Very nice!

jakirkham commented:
An attempt at fixing the style errors, @charris.
